### DISCUSSION TITLE
Elevate `Control.get_rect` to `CanvasItem.get_rect`

### DISCUSSION THREAD

**Discussion by HubbleCommand** [godotengine/godot-proposals#10941] [2D]
As the title says, I'd like to open the discussion to elevating the `get_rect` function of `Control` to `CanvasItem`, so that nodes inheriting from `Node2D`  could implement it as well.

Some Node2Ds like [Sprite2D](https://docs.godotengine.org/en/stable/classes/class_sprite2d.html#class-sprite2d-method-get-rect) do have a public method for it, and other nodes like `Polygon2D` and `Line2D` keep this information internally for editor use.

While in many cases it can be better to work with the Physics server, that can also be excessive. For example, I wanted to get the Rect of a Polygon2D for some basic overlap checks. Would have been overkill to create the resource in the physics server etc. just for a rare check. While it is very easy to determine the Rect in many cases, it would be nice to have it directly from the engine. Even the example for Sprite2D to just determine clicks seems like a handy enough reason to me, albeit not particularly important.

I'm also interested in the reason `get_rect` wasn't made available in `Node2D`, if any. I looked through some discussions, issues, and PRs, and didn't find anything.

This is something I would be willing to work on if it's desired.

**Comment by AThousandShips** [comment]
Not every `CanvasItem` has a well defined rect, just like `Node3D` doesn't expose a `get_aabb` but `VisualInstance3D` does, however every `Control` is supposed to support an area

**Reply by HubbleCommand** [reply]
A Rect is just a position and a size, even if the size is zero for some nodes by their nature (things like Marker, RemoteTransform, or AudioListener). A sizable chunk of Node2Ds do have a size.

I mean, I guess the same thing can be said about Node3D. In either case, I would intuitively assume something that exists in 2D or 3D space to have some size, it feels like the exception that it wouldn't. It's not as abstracted or incorporeal as Node is.

Can I assume that's the reason then? While many do take space, it's that it's not a guarantee?

**Reply by AThousandShips** [reply]
So then you agree that many `Node2D`s don't have a valid `Rect2` distinct from just their position? Just like many `Node3D`s doesn't have a valid `AABB` distinct from their position? Unlike `Control` nodes which _all_ can be said to be _able_ to have a well defined `Rect2`?

**Reply by HubbleCommand** [reply]
Size zero is still valid. We're not talking negative size voodoo.

As long as I know the reason, it's good for me.
